src/runtime/thread_pool.cc

```diff
@@ -127,99 +130,138 @@ class ParallelLauncher {
+      return true;
+    }
+    return false;
   }
+
   /*!
-   * \brief Push task into the queue.
-   * \param task The task to be pushed.
+   * \brief Push a task into the queue and notify the comsumer if it is on wait.
+   * \param input The task to be dequeued.
    */
-  void Push(Task task) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    if (num_pending_ < ring_.size()) {
-      CHECK_NE(ring_.size(), 0U);
-      ring_[(head_ + num_pending_) % ring_.size()] = task;
-      ++num_pending_;
-    } else {
-      size_t old_size = ring_.size();
-      ring_.resize(old_size * 2);
-      if (head_ + num_pending_ > old_size) {
-        // copy the ring overflow part into the tail.
-        size_t ncopy = head_ + num_pending_ - old_size;
-        memcpy(&ring_[0] + old_size, &ring_[0], ncopy * sizeof(Task));
-      }
-      ring_[(head_ + num_pending_) % ring_.size()] = task;
-      ++num_pending_;
+  void Push(const Task& input) {
+    while (!Enqueue(input)) {
+      std::this_thread::yield();
     }
-    if (nwait_consumer_ != 0) {
-      lock.unlock();
+    if (pending_.fetch_add(1) == -1) {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.notify_one();
     }
   }
+
   /*!
-   * \brief Pop task from the queue
-   * \param task The task to be poped.
-   * \param timeout The number of cycles to spin before sleep.
-   * \return Whether pop is successful or we need to exit now.
+   * \brief Pop a task out of the queue and condition wait if no tasks.
+   * \param output The pointer to the task to be dequeued.
+   * \param timeout The number of iterations to spin before sleep.
+   * \return Whether pop is successful (true) or we need to exit now (false).
    */
-  bool Pop(Task* task, int timeout = 10) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    if (num_pending_ != 0) {
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
-    } else {
-      lock.unlock();
-      // do a bit spin and busy waiting before sleep.
-      for (int i = 0; i < timeout && num_pending_ == 0; ++i) {
+  bool Pop(Task* output, uint32_t timeout = 100000) {
+    // busy wait a bit when the queue is empty
+    // if a new task comes to the queue quickly, this wait avoid the worker from sleeping
+    for (uint32_t i = 0; i < timeout && pending_.load() == 0; ++i) {
         std::this_thread::yield();
       }
-      lock.lock();
-      ++nwait_consumer_;
+    if (pending_.fetch_sub(1) == 0) {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.wait(lock, [this] {
-          return num_pending_ != 0 || exit_now_.load();
+          return pending_.load() >= 0 || exit_now_.load();
         });
-      --nwait_consumer_;
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
     }
-    return true;
+    return Dequeue(output);
```

Comment:
Move the Dequeue logic here, here we need to 
```
CHECK(tail_.load(std::memory_order_acquire) != head)
load and return true
```

Reply:
The current logics makes the implicit assumption that queue is not empty, which is fine, but the check would have been more helpful